File: backend/app/models/coin_grant.py
"""Auditoría de coins otorgados por un superadmin — individual (premio a un
jugador puntual) o broadcast (evento para toda la comunidad). `target_user_id`
NULL identifica un broadcast; `recipient_count` solo se completa en ese caso
(cuántos usuarios recibieron el otorgamiento), ya que un broadcast no tiene
un único destinatario que registrar.
"""
import uuid
from datetime import datetime
from typing import Optional

from sqlalchemy import DateTime, ForeignKey, Integer, String, func
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base import Base


class CoinGrant(Base):
    __tablename__ = "coin_grants"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    granted_by_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id"), nullable=False, index=True
    )
    target_user_id: Mapped[Optional[uuid.UUID]] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id"), nullable=True, index=True
    )
    amount: Mapped[int] = mapped_column(Integer, nullable=False)
    reason: Mapped[Optional[str]] = mapped_column(String, nullable=True)
    recipient_count: Mapped[Optional[int]] = mapped_column(Integer, nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )

File: frontend/lib/domain/entities/coin_grant.dart
/// `targetUsername` null identifica un broadcast a toda la comunidad —
/// ahí `recipientCount` cuenta cuántos usuarios lo recibieron.
class CoinGrantEntity {
  final String id;
  final String grantedByUsername;
  final String? targetUsername;
  final int amount;
  final String? reason;
  final int? recipientCount;
  final DateTime createdAt;

  CoinGrantEntity({
    required this.id,
    required this.grantedByUsername,
    this.targetUsername,
    required this.amount,
    this.reason,
    this.recipientCount,
    required this.createdAt,
  });

  bool get isBroadcast => targetUsername == null;

  factory CoinGrantEntity.fromJson(Map<String, dynamic> json) {
    return CoinGrantEntity(
      id: json['id'] as String,
      grantedByUsername: json['granted_by_username'] as String,
      targetUsername: json['target_username'] as String?,
      amount: json['amount'] as int,
      reason: json['reason'] as String?,
      recipientCount: json['recipient_count'] as int?,
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }
}
